vision/odlc_contour_filtering.py

```diff
@@ -0,0 +1,407 @@
+
+Brainstorming:
+Any ODLC shape will have a letter in it so the contour hierarchy will have stuff (needs heirarchy)
+bounding box will probably be close to a square and/or small (def will not be like whole image)
+wont have any crazy points that are really far from center of mass, can use:
+	>>> moms = cv2.moments(single_contour)
+	>>> center_x = (moms["m10"] / moms["m00"])
+	>>> center_y = (moms["m01"] / moms["m00"])
+either will be circular or approxPolyDP will work well min area lost w/o erroneous area incl.
+"""
+from typing import TypeAlias
+import numpy as np
+from nptyping import NDArray, Shape, UInt8, IntC, Float32, Bool
+import cv2
+
+
+# return types for cv2.findContours() -> (Tuple[contours], hierarchy)
+contour_type: TypeAlias = NDArray[Shape["*, 1, 2"], IntC]
+hierarchy_type: TypeAlias = NDArray[Shape["1, *, 4"], IntC]
+
+# type for return value of cv2.boxpoints()
+min_area_box_type: TypeAlias = NDArray[Shape["4, 2"], Float32]
+# type for return value of cv2.boundingRect()
+bound_box_type: TypeAlias = tuple[int, int, int, int]
+# type for return of cv2.HoughCircles() (center_x, center_y, radius)
+circles_type: TypeAlias = NDArray[Shape["1, *, 3"], Float32]
+
+image_type: TypeAlias = NDArray[Shape["*, *, 3"], UInt8]
+# single channel image type
+sc_image_type: TypeAlias = NDArray[Shape["*, *"], UInt8]
+mask_type: TypeAlias = NDArray[Shape["*, *"], Bool]
+
+
+def test_heirarchy(hierarchy: hierarchy_type, contour_index: int) -> bool:
+    """
+    Function to test whether the contour at the given index contains another contour in it
+
+    parameters
+    ----------
+    hierarchy : hierarchy_type
+        The 2nd value returned from cv2.findContours(), describes how contours are inside of
+        other contours
+    contour_index : int
+        The index of the contour to be examined (in the contours tuple and hierarchy array
+        returned from cv2.findContours())
+
+    returns
+    -------
+    True if the given contour has a contour inside of it, False otherwise
+    """
+    if hierarchy[0, contour_index, 2] <= 0:
+        return False
+    return True
+
+
+def test_min_area_box(
+    contour: contour_type, min_box_ratio: float = 0.9, max_box_ratio: float = 1.1
+) -> bool:
+    """
+    will create a box around the given contour that has the smallest possible area
+    (not necessarily upright) and check if the box's aspect ratio is within the given range
+
+    parameters
+    ----------
+    contour : contour_type
+        The individual contour to be evaluated (as returned from cv2.findContours)
+    min_box_ratio : float
+        The lowest acceptable ratio of min area box length to width must be less than the max
+    max_box_ratio : float
+        The highest ratio of min area box length to width acceptable, more than min
+
+    returns
+    -------
+    returns true if the aspect ratio of the min area box is inbetween the min and max
+    """
```

Comment:
Same thing about return variable/type above. Also, note that you should have a return variable listed in the docstring even if you are returning an expression. You don't need to add a variable to your function, just come up with a descriptive name for the value in the docstring.